src/audio/drc/drc_math_generic.c

```diff
@@ -105,16 +123,170 @@ inline int32_t drc_log_fixed(int32_t x)
+ * |	  WordLength| 31		 |	 32		|12.20	  | 3.29		|
+ * |FractionLength| 20		 |	 29		|			  |			|
+ * +--------------+---------+----------+----------+---------+
+ *	 THD+N = 29*6 = -174 dBc
+ * Arguments	 : int32_t thRadFxp Q12.20 input
+ *						int32_t cdcSinTh Q3.29	output
+ * Return Type	 : cdcSinTh
  */
-inline int32_t drc_sin_fixed(int32_t x)
+inline int32_t drc_sin_fixed(int32_t thRadFxp)
 {
-	const int32_t PI_OVER_TWO = Q_CONVERT_FLOAT(1.57079632679489661923f, 30);
-
-	/* input range of sin_fixed() is non-negative */
-	int32_t abs_sin_val = sin_fixed(q_mult(ABS(x), PI_OVER_TWO, 30, 30, 28));
-	return SGN(x) < 0 ? -abs_sin_val : abs_sin_val;
+	/* cordic_atan2_lookup_table = atan(2.^-(0:N-1)) N = 30	*/
+	/* CORDIC Gain is cordic_gain = prod(sqrt(1 + 2.^(-2*(0:30-1)))) */
+	/* Inverse CORDIC Gain,inverse_cordic_gain = 1 / cordic_gain  */
+  static const int32_t inpLUT[30] = { 421657428, 248918915, 131521918, 66762579,
+	 33510843, 16771758, 8387925, 4194219, 2097141, 1048575, 524288, 262144,
+	 131072, 65536, 32768, 16384, 8192, 4096, 2048, 1024, 512, 256, 128, 64, 32,
+	 16, 8, 4, 2, 1 };
+
+  int32_t b_idx;
+  int32_t b_yn;
+  int32_t th;
+  int32_t xn;
+  int32_t xtmp;
+  int32_t ytmp;
+  int32_t z;
+  int32_t cdcSinTh;
+  bool negate;
+
+  th = thRadFxp;
+	 if (th > PIQ20OFFSET0) {											 /* >20bits */
+		if (((th - PIQ21OFFSET0) & PIQ30OFFSET0) != 0) {		 /* NonZero thRadFxp - 21bits & 30bits */
+		  z = (th - PIQ21OFFSET0) | -PIQ30OFFSET0;
+		} else {
+		  z = (th - PIQ21OFFSET0) & PIQ29OFFSET1;
+		}
+
+		if (z <= PIQ20OFFSET0) {												 /*  <20 bits*/
+		  th = z;
+		  negate = true;
+		} else {
+		  if (((th - PIQ22OFFSET0) & PIQ30OFFSET0) != 0) {		 /*  thRadFxp - 22bits & 30bits */
+			 th = (th - PIQ22OFFSET0) | -PIQ30OFFSET0;
+		  } else {
+			 th = (th - PIQ22OFFSET0) & PIQ29OFFSET1;
+		  }
+
+		  negate = false;
+		}
+	 } else if (th < -PIQ20OFFSET0) {									/*	 <20bit */
+		if (((th + PIQ21OFFSET0) & PIQ30OFFSET0) != 0) {			/*	 NonZero thRadFxp - 21bits & 30bits */
+		  z = (th + PIQ21OFFSET0) | -PIQ30OFFSET0;
+		} else {
+		  z = (th + PIQ21OFFSET0) & PIQ29OFFSET1;
+		}
+
+		if (z >= -PIQ20OFFSET0) {
+		  th = z;
+		  negate = true;
+		} else {
+		  if (((th + PIQ22OFFSET0) & PIQ30OFFSET0) != 0) {		 /*  thRadFxp - 22bits & 30bits */
+			 th = (th + PIQ22OFFSET0) | -PIQ30OFFSET0;
+		  } else {
+			 th = (th + PIQ22OFFSET0) & PIQ29OFFSET1;
+		  }
+
+		  negate = false;
+		}
+	 } else {
+		negate = false;
+	 }
+
+	 th <<= 9;															/* thRadFxp *= 512 */
+	 if ((th & PIQ30OFFSET0) != 0) {								/*	 NonZero 30bits */
+		z = th | -PIQ30OFFSET0;
+	 } else {
+		z = th & PIQ29OFFSET1;
+	 }
+
+	 b_yn = 0;
+	 xn = PIQ28OFFSET4;														/*	 28 bits */
+	 xtmp = PIQ28OFFSET4;
+	 ytmp = 0;
+	 for (b_idx = 0; b_idx < 30; b_idx++) {							/*	 Iterate data with lookup table */
+		if (z < 0) {															/*	 negative data */
+		  th = z + inpLUT[b_idx];
+		  if ((th & PIQ30OFFSET0) != 0) {						/*	 30 bits */
```

Comment:
This repeats many times. You could make an inline function with a descriptive function name.

Reply:
Was reviewing earlier in meeting with @ShriramShastry 
I think we can combine ideas
```
static inline int32_t func_name(val, bit1, bit2)
{
    val = val & bit1;
    return th ? th | -bit1 : th & bit2;
}